### PR TITLE
Convert LeNet1989 from Torch to JAX #

### PR DESCRIPTION
**Pull request by peterchang0414**
## Description
JAX/Flax implementation of Andrej Karpathy's [implementation](https://github.com/karpathy/lecun1989-repro) of  [1989 LeNet paper](http://yann.lecun.com/exdb/publis/pdf/lecun-89e.pdf
) using modern methods ([blog post](https://karpathy.github.io/2022/03/14/lecun1989/)).

[Colab notebook](https://colab.research.google.com/github/peterchang0414/lecun1989-flax/blob/main/lecun1989-flax.ipynb)

### Results
**1989 Reproduction**:
```
epoch 23
eval: split train. loss 5.265484e-03. error 0.82%. misses: 60
eval: split test . loss 2.467080e-02. error 3.69%. misses: 74
```
**Modern Versions**:

(1) Replace tanh on last layer with FC  and use softmax. Lower learning rate to 0.01
```
epoch 23
eval: split train. loss 7.162272e-03. error 0.05%. misses: 4
eval: split test . loss 1.687743e-01. error 4.14%. misses: 83
```
(2) Replace SGD with AdamW (learning rate=3e-4), double epochs to 46, and decay LR to 1e-4 over the course of training.
```
epoch 46
eval: split train. loss 1.890260e-03. error 0.04%. misses: 2
eval: split test . loss 1.953933e-01. error 4.04%. misses: 81
```
(3)  Introduce data augmentation, e.g. a shift by at most 1 pixel in both x/y directions, and bump up training time to 60 epochs.
```
epoch 60
eval: split train. loss 5.098452e-02. error 1.65%. misses: 120
eval: split test . loss 9.166716e-02. error 2.59%. misses: 52
```
(4) Add dropout at layer H3, shift activation function to relu, and bring up iterations to 80.
```
epoch 80
eval: split train. loss 3.316079e-02. error 1.06%. misses: 77
eval: split test . loss 4.969697e-02. error 1.74%. misses: 35
```

### Issue 
#698 

## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

### CODE REVIEW

**Comment by peterchang0414** [comment]
I would appreciate a code review: @murphyk @mjsML @gerdm 
Thanks!

**Comment by gerdm** [comment]
Hi @peterchang0414!

I would only add an optional parameter to either store the weights and create the folder or not (with predetermined value `False`). Everything else looks good to me.

**Comment by peterchang0414** [comment]
Thanks for your feedback, @gerdm 

I set the default value for `output_dir=None` and created folder/saved param values only `if args.output_dir`.

**Comment by murphyk** [comment]
I just merged https://github.com/probml/probml-notebooks/blob/main/notebooks/lecun1989_flax.ipynb, so I don't think we need these scripts, do we?
